### PR TITLE
Add `dgNN` support

### PR DESCRIPTION
**Pull request by HenryChang213**
- run `examples/fused_gat.py` (need to install dgNN first, [https://github.com/dgSPARSE/dgNN](url))
- a `FusedGATConv` layer is added in `nn/fused_gat_conv.py`, which is equivalent to the standard GAT model mathmatically with less memory consumption and faster speed, due to kernel fusion techniques, as in paper 'Understanding GNN Computational Graph: A Coordinated Computation, IO, and Memory Perspective'([https://proceedings.mlsys.org/paper/2022/hash/9a1158154dfa42caddbd0694a4e9bdc8-Abstract.html](url))
- a `to_dgnn` converting function is added in `utils/to_dgnn.py` to convert the `edge_index` format used by PyG into the CSR and CSC format used by dgNN

### CODE REVIEW

**Comment by Padarn** [comment]
Hey @HenryChang213 could we add some tests?

File: torch_geometric/nn/conv/fused_gat_conv.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        kwargs.setdefault('aggr', 'add')
+        super().__init__(node_dim=0, **kwargs)
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.heads = heads
+        self.concat = concat
+        self.negative_slope = negative_slope
+        self.dropout = dropout
+        # self.add_self_loops = add_self_loops
```
Comment (Padarn) [inline torch_geometric/nn/conv/fused_gat_conv.py:47]: Maybe want to remove these?

File: torch_geometric/nn/conv/fused_gat_conv.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            assert x.dim() == 2, "Static graphs not supported in 'GATConv'"
+            x_src = x_dst = self.lin_src(x).view(-1, H, C)
+        else:  # Tuple of source and target node features:
+            x_src, x_dst = x
+            assert x_src.dim() == 2, "Static graphs not supported in 'GATConv'"
+            x_src = self.lin_src(x_src).view(-1, H, C)
+            if x_dst is not None:
+                x_dst = self.lin_dst(x_dst).view(-1, H, C)
+
+        x = (x_src, x_dst)
```
Comment (Padarn) [inline torch_geometric/nn/conv/fused_gat_conv.py:102]: This is not used below?

File: torch_geometric/nn/conv/fused_gat_conv.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    def forward(self, x: Union[Tensor, OptPairTensor], edge_index: Adj):
+
+        H, C = self.heads, self.out_channels
+
+        # We first transform the input node features. If a tuple is passed, we
+        # transform source and target node features via separate weights:
+        if isinstance(x, Tensor):
+            assert x.dim() == 2, "Static graphs not supported in 'GATConv'"
+            x_src = x_dst = self.lin_src(x).view(-1, H, C)
+        else:  # Tuple of source and target node features:
```
Comment (Padarn) [inline torch_geometric/nn/conv/fused_gat_conv.py:95]: Curious why you would want to pass a tuple if one of them were `None`? Would it make more sense to have 
`Union[Tensor, PairTensor]` or `OptPairTensor` as the input types?

File: torch_geometric/nn/conv/fused_gat_conv.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
+from typing import Tuple, Union
```
Comment (rusty1s) [inline torch_geometric/nn/conv/fused_gat_conv.py:1]: I think we should be able to inherit from `GATConv` here, and just override its `forward` method. Correct?

File: examples/fused_gat.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
+import argparse
```
Comment (rusty1s) [inline examples/fused_gat.py:1]: Maybe we can add a `benchmark/fuse` repo instead and compare runtimes between `GATConv` and `FusedGATConv`?

File: torch_geometric/utils/to_dgnn.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import scipy.sparse as sp
+import torch
+
+
+def to_dgnn(edge_index):
```
Comment (rusty1s) [inline torch_geometric/utils/to_dgnn.py:5]: Can we add some tests to this function?

File: torch_geometric/nn/conv/fused_gat_conv.py (hunk 4 above)
Comment (fuvty) [reply torch_geometric/nn/conv/fused_gat_conv.py:1]: Yeah, I think so. We can inherit from GATConv instead.

File: torch_geometric/nn/conv/fused_gat_conv.py (hunk 2 above)
Comment (fuvty) [reply torch_geometric/nn/conv/fused_gat_conv.py]: Yeah. Thanks for pointing it out, we deleted it.

File: torch_geometric/utils/to_dgnn.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
```
Comment (rusty1s) [inline torch_geometric/utils/to_dgnn.py:5]: Would be also good to name it more generally, e.g., `compute_all_graph_representations`.

File: torch_geometric/utils/to_dgnn.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    device = edge_index.device
+
+    col, row = edge_index.cpu()
+    numlist = torch.arange(col.size(0), dtype=torch.int32)
+    num_nodes = torch.max(col) - torch.min(col) + 1
+
+    adj_csr = sp.csr_matrix((numlist.numpy(), (row, col)),
```
Comment (rusty1s) [inline torch_geometric/utils/to_dgnn.py:12]: Can we implement this via PyTorch manually? Current implementation is CPU-only :(

File: torch_geometric/nn/conv/fused_gat_conv.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
+from typing import Optional, Tuple, Union
+
+from dgNN.operators import GATConvFuse
```
Comment (rusty1s) [inline torch_geometric/nn/conv/fused_gat_conv.py:3]: Can we make this an optional dependency? Similar to [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/spline_conv.py#L15)?

**Comment by MathsCode** [comment]
The test errors in the above check are the same ('Trainer' object has no attribute 'lr_schedulers' in test_graphgym.py). The test_graphgym.py uses the class "LoggerCallback()" in line 187. And the LoggerCallback() is in logger.py (torch_geomtric/graphgym/logger.py). There is a function called _get_stats using the "lr_schedulers". I think this is the root of the problem.

**Comment by fuvty** [comment]
> The test errors in the above check are the same ('Trainer' object has no attribute 'lr_schedulers' in test_graphgym.py). The test_graphgym.py uses the class "LoggerCallback()" in line 187. And the LoggerCallback() is in logger.py (torch_geomtric/graphgym/logger.py). There is a function called _get_stats using the "lr_schedulers". I think this is the root of the problem.

It is caused by another recent commit to the master branch of PyG, so we did not fix this issue in this PR. All the tests related to this PR have passed :)

**Comment by fuvty** [comment]
The code looks good to me now :)

**Review by rusty1s — APPROVED** [review]

**Pull request merged**